### PR TITLE
Always evaluate free constants and statics, even if previous errors occurred

### PR DESCRIPTION
**Pull request by oli-obk**
work towards https://github.com/rust-lang/rust/issues/79738

We will need to evaluate static items before the `definitions.freeze()` below, as we will start creating new `DefId`s (for nested allocations) within the `eval_static_initializer` query.

But even without that motivation, this is a good change. Hard errors should always be reported and not silenced if other errors happened earlier.

### CODE REVIEW

File: compiler/rustc_lint/src/builtin.rs [hunk 1]
```diff
@@ -1542,32 +1542,6 @@ impl<'tcx> LateLintPass<'tcx> for TypeAliasBounds {
...
-            }
-            hir::ItemKind::Static(_, _, body_id) => {
-                let def_id = cx.tcx.hir().body_owner_def_id(body_id).to_def_id();
-                cx.tcx.ensure().eval_static_initializer(def_id);
-            }
-            _ => {}
-        }
-    }
-}
-
```
Comment (lcnr) [inline compiler/rustc_lint/src/builtin.rs:1570]: that was a lint before :sweat_smile:

File: compiler/rustc_middle/src/mir/interpret/queries.rs [hunk 2]
```diff
@@ -192,6 +192,15 @@ impl<'tcx> TyCtxtAt<'tcx> {
         self,
         def_id: DefId,
     ) -> Result<mir::ConstAllocation<'tcx>, ErrorHandled> {
+        if let Some(def_id) = def_id.as_local() {
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/queries.rs:195]: Please add a comment explaining what this is about.

File: compiler/rustc_hir_analysis/src/lib.rs [hunk 3]
```diff
@@ -196,6 +196,15 @@ pub fn check_crate(tcx: TyCtxt<'_>) -> Result<(), ErrorGuaranteed> {
         collect::test_opaque_hidden_types(tcx)?;
     }
 
+    tcx.hir().par_body_owners(|item_def_id| {
```
Comment (RalfJung) [inline compiler/rustc_hir_analysis/src/lib.rs:199]:
```suggestion
    // Make sure we evaluate all static and (non-associated) const items, even if unused.
    // If any of these fail to evaluate, we do not want this crate to pass compilation.
    tcx.hir().par_body_owners(|item_def_id| {
```
And maybe also say something about the DefId generation, which forces this to be here? I didn't understand that part.

File: tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr [hunk 4]
```diff
@@ -4,8 +4,14 @@ error[E0764]: mutable references are not allowed in the final value of constants
 LL | const B: *mut i32 = &mut 4;
    |                     ^^^^^^
 
+error: encountered mutable pointer in final value of constant
+  --> $DIR/mut_ref_in_final.rs:10:1
+   |
+LL | const B: *mut i32 = &mut 4;
+   | ^^^^^^^^^^^^^^^^^
```
Comment (RalfJung) [inline tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr:11]: This error is basically a duplicate of the one above. Why do we get both of them now?

File: compiler/rustc_hir_analysis/src/lib.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/lib.rs:199]: That is a comment I already rebased away locally 😅

Comment (RalfJung) [reply compiler/rustc_hir_analysis/src/lib.rs:199]: I... don't know what that means.^^

File: tests/ui/consts/miri_unleashed/mutable_references.rs [hunk 5]
```diff
@@ -4,29 +4,27 @@ use std::cell::UnsafeCell;
...
 static BAR: &mut () = &mut ();
+//~^ ERROR encountered mutable pointer in final value of static
 
 struct Foo<T>(T);
 
-// this is fine for the same reason as `BAR`.
 static BOO: &mut Foo<()> = &mut Foo(());
+//~^ ERROR encountered mutable pointer in final value of static
 
 // interior mutability is fine
```
Comment (lcnr) [inline tests/ui/consts/miri_unleashed/mutable_references.rs:18]: outdated comment, this test generally feels somewhat useless rn, given that its behavior doesn't depend on `-Zunleash-the-miri-inside-of-you` afaict?

File: tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr (hunk 4 above)
Comment (oli-obk) [reply tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr:11]: One is  const checking failure, the other is an evaluation failure. I'll try to make const eval not run if checking failed

**Review by lcnr — COMMENTED** [review]
r=me after nits + CI

File: compiler/rustc_lint/src/builtin.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_lint/src/builtin.rs:1570]: Yes.^^ I'm so happy that this hack is gone.

(To be clear, the lint was triggering hard errors. It just used the lint machinery to get "traverse the entire crate" for free.)

File: tests/ui/consts/miri_unleashed/mutable_references.rs (hunk 5 above)
Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references.rs:18]: stderr has a bunch of "skipping check that does not even have a feature gate", so I think it does depend on the unleash flag.

File: compiler/rustc_hir_analysis/src/lib.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_hir_analysis/src/lib.rs:199]: > And maybe also say something about the DefId generation

that's not happening. We're not generating def ids yet. I will do so in the future and then update the comment here

File: compiler/rustc_const_eval/src/transform/check_consts/check.rs [hunk 6]
```diff
@@ -249,7 +249,7 @@ impl<'mir, 'tcx> Checker<'mir, 'tcx> {
         let secondary_errors = mem::take(&mut self.secondary_errors);
         if self.error_emitted.is_none() {
             for error in secondary_errors {
-                error.emit();
+                self.error_emitted = Some(error.emit());
```
Comment (oli-obk) [inline compiler/rustc_const_eval/src/transform/check_consts/check.rs:252]: @RalfJung found the issue that caused us to const eval even though there were const qualif errors 🙈

**Comment by bjorn3** [comment]
> as we will start creating new DefIds (for nested allocations) within the eval_static_initializer query.

Wouldn't it be possible to represent the id of nested statics as a pair of "def id of containing static" and "index of this nested static"?

**Comment by oli-obk** [comment]
> Wouldn't it be possible to represent the id of nested statics as a pair of "def id of containing static" and "index of this nested static"?

yes, but that just makes everything more complex, as then everything needs to handle this special case. I tried it, and it's annoying to get right. Just generating a new static per nested allocation was significantly simpler.

**Comment by cjgillot** [comment]
This has conflicts.
@bors r-

**Comment by oli-obk** [comment]
Looking into the regressions. They are unintended

**Comment by oli-obk** [comment]
@rustbot label: +perf-regression-triaged

I fixed the libc regression in #121387 

The other regressions' detailed view actually shows improvements instead of a regression, and the benchmarks themselves have been flaky on the same scale as they were affected here, so I'm guessing it's not related

**Pull request merged**